src/Language/PureScript/CoreImp/Optimizer/MagicDo.hs

```diff
@@ -73,6 +73,8 @@ magicDo = inlineST . everywhere undo . everywhereTopDown convert
   -- Remove __do function applications which remain after desugaring
   undo :: AST -> AST
   undo (Return _ (App _ (Function _ (Just ident) [] body) [])) | ident == fnName = body
+  undo (App _ (App s1 (Function s2 Nothing [] (Block ss body)) []) []) =
```

Comment:
Defer to @natefaubion here. But the example code is in https://github.com/purescript/purescript/pull/3026#issuecomment-320511849 - with this change an extra thunk is removed.